File: core/src/net/natruid/jungle/systems/FlowControlSystem.java
package net.natruid.jungle.systems;

import com.artemis.Aspect;
import com.artemis.ComponentMapper;
import net.natruid.jungle.components.render.ActingComponent;
import net.natruid.jungle.systems.abstracts.PassiveEntitySystem;

public class FlowControlSystem extends PassiveEntitySystem {

    private ComponentMapper<ActingComponent> mActing;

    public FlowControlSystem() {
        super(Aspect.all(ActingComponent.class));
    }

    public boolean isReady() {
        return getEntityIds().isEmpty();
    }

    public void addAct(int entityId) {
        mActing.create(entityId).actions += 1;
    }

    public void delAct(int entityId) {
        ActingComponent actingComponent = mActing.get(entityId);
        actingComponent.actions -= 1;
        if (actingComponent.actions == 0) {
            mActing.remove(entityId);
        }
    }
}

File: core/src/net/natruid/jungle/systems/FlowControlSystem.kt
package net.natruid.jungle.systems

import com.artemis.Aspect
import com.artemis.BaseEntitySystem
import com.artemis.ComponentMapper
import net.natruid.jungle.components.render.ActingComponent

class FlowControlSystem : BaseEntitySystem(Aspect.all(ActingComponent::class.java)) {

    val ready get() = entityIds.isEmpty

    private lateinit var mActing: ComponentMapper<ActingComponent>

    override fun processSystem() {}

    fun addAct(entityId: Int) {
        mActing.create(entityId).apply {
            actions += 1
        }
    }

    fun delAct(entityId: Int) {
        mActing[entityId].apply {
            actions -= 1
            if (actions == 0) {
                mActing.remove(entityId)
            }
        }
    }
}
